### PR TITLE
Fix Issue 14874 - __traits(getFunctionAttributes) does not support the new `return` attribute

### PR DESCRIPTION
**Pull request by MetaLang**
https://issues.dlang.org/show_bug.cgi?id=14874

### CODE REVIEW

**Comment by WalterBright** [comment]
need to compile with `-dip25` switch

**Comment by MetaLang** [comment]
I wondered why the autotester was choking. Do I just have to add a `REQUIRED_ARGS: dip25` to the top of the test file?

File: test/runnable/test14874.d [hunk 1]
```diff
@@ -0,0 +1,36 @@
+template indexOfReturn(T...)
+{
+    static if (T.length == 0)
+    {
+        return -1;
```
Comment (9rnsr) [inline test/runnable/test14874.d]: `return` statement  in template member scope is invalid.

Comment (MetaLang) [inline test/runnable/test14874.d]: Oh, whoops, I thought I got rid of that. I'll have to fix it up in a bit.

Comment (9rnsr) [inline test/runnable/test14874.d]: And, you can add the test code in the test file `runnable\traits.d`. Correcting `__traits()` related test cases into that source file would be better.

File: test/runnable/test14874.d [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    ref int getNNonReturn()
+    {
+        return n;
+    }
+}
+
+void main()
+{
+    assert(indexOfReturn!(__traits(getFunctionAttributes, Test.getN)) != -1);
+    assert(indexOfReturn!(__traits(getFunctionAttributes, Test.getNNonReturn)) != -1);
```
Comment (9rnsr) [inline test/runnable/test14874.d]: Instead using `static assert` will test the result at compile time.

Comment (9rnsr) [inline test/runnable/test14874.d]: Perhaps the last line is: `== -1`?

Comment (MetaLang) [inline test/runnable/test14874.d]: Yeah that should `==`.

**Comment by MetaLang** [comment]
Okay, we should be in good shape now. I won't be able to squash the extra commits until later today, however.

**Comment by MetaLang** [comment]
> core.exception.AssertError@runnable/testscope2.d(45): Assertion failure

Is this related or just a freak error?

**Comment by 9rnsr** [comment]
You have 4 commits for this PR, so to squack the commits, you can use

`git rebase -i head~4`

then let's push the new one commit forcibly to your remote.

**Comment by 9rnsr** [comment]
> Is this related or just a freak error?

It's related. In `hdrgen.c` there are hacky `STCreturn` handlings for the function type printing around line 880 and 945. By them, in the test case `runnable/testscope2.d` `stringof` property prints `return` attribute twice. Your `TypeFunction::attributesApply` fix will make the two hacks in `hdrgen.c` unnecessary. Let's remove them!

**Comment by MetaLang** [comment]
Sounds good to me. I'll do that sometime this evening.

**Comment by MetaLang** [comment]
Also, is there a way to run the DMD tests locally? It's probably better than abusing the auto-tester.

**Comment by 9rnsr** [comment]
You can read the information in wiki page: http://wiki.dlang.org/Building_DMD#Running_the_test_suite

**Comment by 9rnsr** [comment]
The final error in test case is the line 45 in `runnable/testscope2.d`:

```d
	assert(typeof(SS.foo1).stringof == "ref int(return ref int delegate() return p)return ");
```

The type printing relying on `TypeFunction::attributesApply` will fix the incorrect spacing, then it will be changed to:

```d
	assert(typeof(SS.foo1).stringof == "ref int(return ref int delegate() return p) return");
```

**Comment by 9rnsr** [comment]
One more suggestion is, let's format the first line of commit message to "fix Issue xxxxx - title". When bugzilla receives a notification mail from github, it will parse that specific message format, then comment the commit log to issue xxxxx ticket. it will add traceability to the actual commit which fixing the bug.

**Comment by 9rnsr** [comment]
Oh, sorry... The correct output is: `ref return int(return ref int delegate() return p)`. I tested this PR in my local, then I confirmed it.

That looks change the position of `return` attribute, but when a function type (== `TypeFunction`) is printed, all other function attributes (ref, const, immutable, pure, @safe, etc) are formatted by using prefix style. So the change will increase consistency. I think it's good.

**Comment by MetaLang** [comment]
So the version with the outermost `return` as a prefix is correct?

**Comment by 9rnsr** [comment]
Yes.

...And, the finished compiler bugfix seems breaking std.traits unittest.

**Comment by MetaLang** [comment]
I have opened a Phobos PR that fixes the unittests in std.traits.

https://github.com/D-Programming-Language/phobos/pull/3547

**Comment by 9rnsr** [comment]
Auto-merge toggled on

**Comment by MetaLang** [comment]
This fix did not get cherry-picked into the stable branch for 2.068. Was that an oversight? I was hoping not to have to wait for 2.069 for my code to compile.

**Comment by 9rnsr** [comment]
`stable` for 2.068.x release had branched [since 6/18](https://github.com/D-Programming-Language/dmd/tree/v2.068.0-b1).

2.068 is released under the [DIP75 process](http://wiki.dlang.org/DIP75). By that, bugfix commits in master branch don't get cherry-picked into stable branch.

**Comment by MetaLang** [comment]
Oh, you're right, 2.068 was out far before this. Hmm, it doesn't seem that long ago. My mistake.

**Pull request merged**